pkg/tabulator/tabstate.go

```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
 	}(ctx)
 
 	// Set up threads
-	var active stringset.Set
-	var waiting stringset.Set
-	var lock sync.Mutex
+	var activeGroup stringset.Set
+	var activeGroupLock sync.Mutex
+	groups := make(chan *configpb.TestGroup)
+
+	tabs := make(chan writeTask)
+	var tabLock sync.Mutex
+
+	read := func(ctx context.Context, log *logrus.Entry, group *configpb.TestGroup) error {
+		activeGroupLock.Lock()
+		ok := activeGroup.Add(group.Name)
+		activeGroupLock.Unlock()
+		if !ok {
+			q.Fix(group.Name, time.Now().Add(freq/10), false)
+			return nil
+		}
+
+		fromPath, err := updater.TestGroupPath(configPath, gridPathPrefix, group.Name)
+		if err != nil {
+			return fmt.Errorf("can't make tg path %q: %w", group.Name, err)
+		}
+
+		log.WithFields(logrus.Fields{
+			"from": fromPath.String(),
+		}).Info("Reading state")
 
-	channel := make(chan *configpb.TestGroup)
+		grid, _, err := gcs.DownloadGrid(ctx, client, *fromPath)
+		if err != nil {
+			return fmt.Errorf("downloadGrid(%s): %w", fromPath, err)
+		}
 
-	update := func(log *logrus.Entry, group *configpb.TestGroup) error {
+		tabLock.Lock()
+		// lock out all other readers so that all these tabs get handled as soon as possible
+		var wg sync.WaitGroup
+		wg.Add(len(groupToTabs[group.Name]))
 		for _, view := range groupToTabs[group.Name] {
-			log := log.WithField("dashboard", view.dashboard.Name).WithField("tab", view.tab.Name)
-			fromPath, err := updater.TestGroupPath(configPath, gridPathPrefix, view.tab.TestGroupName)
-			if err != nil {
-				return fmt.Errorf("can't make tg path %q: %w", view.tab.TestGroupName, err)
-			}
-			toPath, err := TabStatePath(configPath, tabsPathPrefix, view.dashboard.Name, view.tab.Name)
-			if err != nil {
-				return fmt.Errorf("can't make dashtab path %s/%s: %w", view.dashboard.Name, view.tab.Name, err)
-			}
-			log.WithFields(logrus.Fields{
-				"from": fromPath.String(),
-				"to":   toPath.String(),
-			}).Info("Calculating state")
-
-			err = createTabState(ctx, log, client, view.tab, group, *fromPath, *toPath, confirm, dropEmptyCols, calculateStats, useTabAlertSettings)
-			if err != nil {
-				return fmt.Errorf("can't calculate state: %w", err)
-			}
+			out := view
+			out.data = proto.Clone(grid).(*statepb.Grid)
+			out.done = wg.Done
```

Comment:
hmm, are you sure about this? Seems fragile

Reply:
WaitGroup seems fine, but passing a wait group to a foreign function via a channel seems fragile and hard to reason about.

Reply:
(I want be able to read a single function to validate that `wg.Done()` is called `wg.Add(n)` times)

Reply:
Maybe instead of passing a wg.Done why not have the writer remove the item from the processing set?

Reply:
The writer might not be the only one writing a test group; test groups to tabs are one-to-many. The processing set could be dashboard#tabs instead, though. 🤔

Reply:
Happy to go with this and wait until there's empirical evidence from debugging a mysterious "why does the tabulator stop doing anything after awhile?" issue after a refactor to conclude that this is a fragile pattern.

Reply:
I thought a bit more about _why_ we want to call back to the reading thread like this at all, and I'm not convinced we need it. These routines are already architected to pass off everything before another reader gets to pass off anything.

Deleting code is a pretty good way to make it simpler, so I would favor that solution